tests/scripts/keadm_compatibility_e2e.sh

```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/env bash
+
+# Copyright 2023 The KubeEdge Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+KUBEEDGE_ROOT=$PWD
+WORKDIR=$(dirname $0)
+E2E_DIR=$(realpath $(dirname $0)/..)
+IMAGE_TAG=$(git describe --tags)
+CLOUD_EDGE_VERSION=${1:-"v1.15.1"}
+
+
+source "${KUBEEDGE_ROOT}/hack/lib/install.sh"
+
+function cleanup() {
+  sudo pkill edgecore || true
+  helm uninstall cloudcore -n kubeedge && kubectl delete ns kubeedge  || true
+  kind delete cluster --name test
+  sudo rm -rf /var/log/kubeedge /etc/kubeedge /etc/systemd/system/edgecore.service $E2E_DIR/e2e_keadm/e2e_keadm.test $E2E_DIR/config.json
+}
+
+function build_ginkgo() {
+  cd $E2E_DIR
+  ginkgo build -r e2e_keadm/
+}
+
+function prepare_cluster() {
+  kind create cluster --name test
+
+  echo "wait the control-plane ready..."
+  kubectl wait --for=condition=Ready node/test-control-plane --timeout=60s
+
+  kubectl create clusterrolebinding system:anonymous --clusterrole=cluster-admin --user=system:anonymous
+
+  # edge side don't support kind cni now, delete kind cni plugin for workaround
+  kubectl delete daemonset kindnet -nkube-system
+}
+
+function build_image() {
+  cd $KUBEEDGE_ROOT
+  make image WHAT=installation-package -f $KUBEEDGE_ROOT/Makefile
+  docker save kubeedge/installation-package:$IMAGE_TAG > installation-package.tar
+  sudo ctr -n=k8s.io image import installation-package.tar
+  kind load docker-image docker.io/kubeedge/installation-package:$IMAGE_TAG --name test
+
+  set +e
+  docker rmi $(docker images -f "dangling=true" -q)
+  set -Ee
+}
+
+function get_cloudcore_image() {
+   docker pull kubeedge/cloudcore:$CLOUD_EDGE_VERSION
+   docker tag kubeedge/cloudcore:$CLOUD_EDGE_VERSION docker.io/kubeedge/cloudcore:$CLOUD_EDGE_VERSION
+   kind load docker-image docker.io/kubeedge/cloudcore:$CLOUD_EDGE_VERSION --name test
+
+   set +e
+   docker rmi $(docker images -f "dangling=true" -q)
+   docker system prune -f
+   set -Ee
+
+}
+
+function get_edgecore_image() {
+   docker run --rm kubeedge/installation-package:$CLOUD_EDGE_VERSION cat /usr/local/bin/keadm > /usr/local/bin/edgeadm && chmod +x /usr/local/bin/edgeadm
```

Comment:
Could you please assist me with understanding this code? Why should you use edgeadm join instead than keadm join?

Reply:
done